lib/omniauth/auth0/jwt_validator.rb

```diff
@@ -0,0 +1,122 @@
+require 'base64'
+require 'uri'
+require 'json'
+require 'omniauth'
+
+module OmniAuth
+  module Auth0
+    class JWTValidator
+
+      attr_accessor :issuer, :jwks
+
+      # Initializer
+      # @param options class
+      #   options.domain - Application domain.
+      #   options.client_id - Application Client ID.
+      #   options.client_secret - Application Client Secret.
+      def initialize(options)
+        @domain = options.domain
+        @issuer = "#{domain_url}/"
+        @client_id = options.client_id
+        @client_secret = options.client_secret
+        @jwks = nil
+      end
+
+      # Decode a JWT.
+      # @param jwt string - JWT to decode.
+      # @return hash - The decoded token, if there were no exceptions.
+      # @see https://github.com/jwt/ruby-jwt
+      def decode(jwt)
+        head = token_head(jwt)
+
+        # Make sure the algorithm is supported and get the decode key.
+        if head[:alg] == 'RS256'
+          jwks_x5c = jwks_key(:x5c, head[:kid])
+          raise JWT::VerificationError, :jwks_missing_x5c if jwks_x5c.nil?
+          decode_key = jwks_public_cert(jwks_x5c.first)
+        elsif head[:alg] == 'HS256'
+          decode_key = @client_secret
+        else
+          raise JWT::VerificationError, :id_token_alg_unsupported
+        end
+
+        # Docs: https://github.com/jwt/ruby-jwt#add-custom-header-fields
+        decode_options = {
+            algorithm: head[:alg],
+            leeway: 30,
+            verify_expiration: true,
+            verify_iss: true,
+            iss: @issuer,
+            verify_aud: true,
+            aud: @client_id,
+            verify_not_before: true
+        }
+
+        # Docs: https://github.com/jwt/ruby-jwt#algorithms-and-usage
+        JWT.decode(jwt, decode_key, true, decode_options)
+      end
+
+      # Get the decoded head segment from a JWT.
+      # @return hash - The parsed head of the JWT passed, empty hash if not.
+      def token_head(jwt)
+        jwt_parts = jwt.split('.')
+        return {} if blank?(jwt_parts) || blank?(jwt_parts[0])
```

Comment:
> If you return {} from here, can the rest of the library execute?

The part of this lib that uses it will fail on looking for an algorithm, which seems rational to me for an empty head. 

> What would you name this line if you had to give it a name?

Not sure what you're asking here. 

> What are we checking for?

Whether or not was have something to parse.